File: scala/17.scala
object Solution {
    def letterCombinations(digits: String): List[String] = {
      val m = Map(
        '2' -> "abc",
        '3' -> "def",
        '4' -> "ghi",
        '5' -> "jkl",
        '6' -> "mno",
        '7' -> "pqrs",
        '8' -> "tuv",
        '9' -> "wxyz"
      )
      def gen(digits: String): List[String] = {
        if (digits.isEmpty()) return Nil
        val res = gen(digits.substring(1)) match {
            case Nil => List("")
            case default => default
        }
        m(digits.head).flatMap(ch => res.map(str => ch + str)).toList
      }
      gen(digits)
    }
}

File: go/17.go
var letterMap = map[byte]string{
	'2': "abc",
	'3': "def",
	'4': "ghi",
	'5': "jkl",
	'6': "mno",
	'7': "pqrs",
	'8': "tuv",
	'9': "wxyz",
}

func letterCombinations(digits string) []string {
	if digits == "" {
		return []string{}
	}
	v := letterCombinations(digits[1:])
	if len(v) == 0 {
		for _, ch := range letterMap[digits[0]] {
			v = append(v, string(ch))
		}
		return v
	}
	var res []string
	for _, ch := range letterMap[digits[0]] {
		for i := range v {
			res = append(res, string(ch)+v[i])
		}
	}
	return res
}